In [1]:
import bs4
from bs4 import BeautifulSoup
import requests
import pandas as pd
import numpy as np
df = pd.read_pickle('russell3000_list_v4.pkl')

# Features we want to create:

### 1. Earnings per share (EPS)
This is the amount each share
 would get if a company paid out all of its profit to its shareholders. EPS is calculated by dividing the company’s total profit by the number of shares.

Example – If a company’s profit is $200 million and there are 10 million shares, the EPS is $20.

EPS can tell you how companies in the same industry compare. Companies that show steady, consistent earnings
 growth, year after year, will often outperform companies with volatile earnings over time.



In [2]:
'''
Questions for Joan:

1. How can I approach lots of null values in the dataset?
2. is 3000 sample size big enough?
3. how to make scraping go faster? - 30 minutes per value scraped
4. How many features should I try to scrape in?


'''

'\nQuestions for Joan:\n\n1. How can I approach lots of null values in the dataset?\n2. is 3000 sample size big enough?\n3. how to make scraping go faster? - 30 minutes per value scraped\n4. How many features should I try to scrape in?\n\n\n'

In [3]:
df

Company Ticker Open_Price_2018 Open_Price_2020  \
0            1-800 FLOWERS COM   FLWS           11.15           13.96   
1     1ST CONSTITUTION BANCORP   FCCY            18.2           22.35   
2              1ST SOURCE CORP   SRCE           49.89           51.25   
3       22ND CENTURY GROUP INC   XXII            3.12            1.13   
4                       2U INC   TWOU           64.26           23.12   
...                        ...    ...             ...             ...   
3007                ZUMIEZ INC   ZUMZ            23.1           33.75   
3008                     ZUORA    ZUO             NaN           14.56   
3009   ZYNERBA PHARMACEUTICALS   ZYNE           12.81             5.8   
3010                     ZYNEX   ZYXI            3.24            7.81   
3011                 ZYNGA INC   ZNGA            3.92            6.12   

      Percent Increase  
0            25.201794  
1            22.802198  
2             2.725997  
3           -63.782051  
4           -64.021164  
...                ...  
3007         46.103896  
3008               NaN  
3009        -54.722873  
3010        141.049383  
3011         56.122449  

[3012 rows x 5 columns]

In [4]:
def get_EPS(row):
    ''' 
    This method scrapes Yahoo Finance for EPS data on Jan 3rd 2020
    '''
    url = 'https://finance.yahoo.com/quote/{stock}/financials?p={stock}'.format(stock = row)
    response = requests.get(url)
    page = response.text
    soup = BeautifulSoup(page, 'lxml')
    try:
        tag = soup.find('span', string = 'Basic EPS')
        eps_2017 = tag.findNext('span').findNext('span').findNext('span')
        print(eps_2017.text)
        return eps_2017.text
    except:
        print(np.NaN)
        return np.NaN

    #         for i,tag in enumerate(tags):
#             if tags[i].text == 'Basic EPS':
#                 EPS_2017 = tags[i+3].text
#                 print(EPS_2017)
#                 return EPS_2017
    #try to calculate EPS
    #what does an EPS of 0 mean?

In [5]:
%%time
df['EPS'] = df['Ticker'].apply(get_EPS)
df

0
0
0.0026
0
0
0
0.0081
0
0
0.001
0.0017
0.0041
0
0.0033
0
0
0.0012
0
0
0
0
0.0023
-0.0024
0
-0.0012
-0.0027
0.0056
0.0012
0
-0.0011
-0.0038
nan
0
-0.0024
0.0015
-0.0049
0
nan
0.0075
0.0012
-0.004
0.001
0
0.0094
-0.0019
0.0034
0
0.0019
0.0012
nan
-0.0014
0
0.0047
0.0157
0.0016
0.002
-0.0011
0.0029
-0.001
-0.001
0.0058
0
0.0013
0.0119
0.0029
0.0382
0.0034
0.0142
0.002
0
0.0034
Diluted EPS
-0.0011
0.0085
0.002
-0.0054
0
0.0183
0.0183
0
-0.0019
0
0.0022
0.017
0.0018
0.0053
0
-0.0057
0
0.0063
-0.0073
0
nan
-0.0038
0.0073
0.003
0
0
0.0015
0
0
0.0022
0.0032
0.0014
0.003
0
0
0.002
0.015
0
0.0047
0.0034
0.002
-0.0037
0.0011
0.0028
0.0015
0
0
0
0
0.0093
0.0034
0.0032
0.0039
0.0035
0.0051
0
0.0014
nan
0
0.0015
nan
0.0023
0.0014
0.0101
0.0029
0
0.0019
0.0018
-0.0022
0
-0.0014
0
-0.0028
0.0011
-0.0056
-0.0026
0.0039
0.0047
0.0451
-0.0011
0
0.0064
0.0012
0.0017
-0.003
nan
0.0032
0
0
0.0043
0.0018
0.0041
-0.0027
0
0.0015
0
0.0041
-0.0019
-0.0049
0
0
0
0.0021
-0.0013
0
0.0017
0.0033
0
0.0012
nan
0.00

-0.0012
0
0.0021
0
-0.0068
0.002
0.0021
-0.0026
0.0073
0.0138
0
0.0019
0
0
0.0013
-0.0031
-0.0018
0
0
0.0038
0
0.0078
0
0
-0.0032
0.0013
0.0012
0.0203
0.0022
0
0.0039
0
0.003
0
0.0012
0.0039
0.002
0.003
0
0.0054
0
-0.0065
0.0183
0.0018
0.0023
0
0
0
0.0019
-0.002
0.0027
0
0.0044
0.0024
0.0026
0
0.0096
0.002
0.0024
0.0019
0.0015
0.003
0.0027
-0.0019
0.0011
0.0028
0
0
0.0021
-0.003
nan
0.0021
0
-0.0015
0
0.0046
0.0015
0.0058
0
0
0
0.0042
-0.0015
0.0029
0
nan
0.0045
0.002
0
0.0015
0
-0.0014
0.0067
-0.1096
0.003
-0.0013
0
0.0021
-0.0024
-0.0016
0.0015
0
0.0095
0.0061
0.0016
0
0
0
0.0048
-0.004
0.0074
-0.0023
0
0.0013
0
0
0.0017
0.0089
0
0.0037
0.002
0
0.0198
0.0037
0.0028
0.002
0.0034
0.0015
0
0.0056
0.0018
0.0019
0.0017
0
nan
0.0011
0.0035
0.0017
0.0013
0
nan
0
0
0.0017
0.001
0.0047
0.003
0
0.0014
0
0
0.0029
nan
0.0014
0.0027
-0.0158
0.0033
0
-0.0044
0
0
0.0044
0.0408
0
0
0.0119
-0.005
0
0
0.0039
nan
0.0021
0.0027
0
-0.0032
0
0
0.003
0
0
0
0.0044
0.0028
-0.0031
nan
0
0
0.0014
nan
nan
0
0
0

Company Ticker Open_Price_2018 Open_Price_2020  \
0            1-800 FLOWERS COM   FLWS           11.15           13.96   
1     1ST CONSTITUTION BANCORP   FCCY            18.2           22.35   
2              1ST SOURCE CORP   SRCE           49.89           51.25   
3       22ND CENTURY GROUP INC   XXII            3.12            1.13   
4                       2U INC   TWOU           64.26           23.12   
...                        ...    ...             ...             ...   
3007                ZUMIEZ INC   ZUMZ            23.1           33.75   
3008                     ZUORA    ZUO             NaN           14.56   
3009   ZYNERBA PHARMACEUTICALS   ZYNE           12.81             5.8   
3010                     ZYNEX   ZYXI            3.24            7.81   
3011                 ZYNGA INC   ZNGA            3.92            6.12   

      Percent Increase      EPS  
0            25.201794        0  
1            22.802198        0  
2             2.725997   0.0026  
3           -63.782051        0  
4           -64.021164        0  
...                ...      ...  
3007         46.103896   0.0011  
3008               NaN  -0.0011  
3009        -54.722873  -0.0025  
3010        141.049383        0  
3011         56.122449        0  

[3012 rows x 6 columns]

In [6]:
df.to_pickle('eps_data.pkl')


In [7]:
import pandas as pd
df = pd.read_pickle('eps_data.pkl')
def convert_eps_to_float(row):
    try:
        return (float(row)*1000)
    except:
        return np.NaN
df['EPS'] = df['EPS'].apply(convert_eps_to_float)
df['EPS'].value_counts()

0.0       1005
1.4         61
1.3         57
2.1         54
1.5         53
          ... 
14.3         1
4792.2       1
11.6         1
14.9         1
28.3         1
Name: EPS, Length: 253, dtype: int64

### 2. Price to earnings (P/E) ratio
This measures the relationship between the earnings of a company and its stock
 price. It’s calculated by dividing the current price per share of a company’s stock by the company’s earnings per share.

Example – A company’s stock currently sells for $50 per share and its earnings per share
 are $5. That means it has a P/E ratio of 10 ($50 divided by $5).

The P/E ratio can tell you whether a stock’s price is high, or low, compared to its earnings.

Some investors consider a company with a high P/E to be overpriced. But sometimes a company with a high P/E today may offer higher returns, and a better P/E, in the future. How do you know? You’ll likely have to look at other indicators before you decide.

In [8]:
import pandas as pd
import numpy as np
def get_PE_ratio(row):
    try:
        return row['Open_Price_2018']/float(row['EPS'])
    except:
        return np.NaN
df['P/E Ratio'] = df.apply(get_PE_ratio, axis = 1)

In [9]:
df['P/E Ratio'].value_counts(dropna = False)

 NaN          1222
 16.700000       3
 23.000000       3
 17.750000       2
 22.153846       2
              ... 
 22.083333       1
-0.946262        1
 34.320000       1
 16.836364       1
-23.450704       1
Name: P/E Ratio, Length: 1768, dtype: int64

In [10]:
df.to_pickle('PE_Ratio_Data.pkl')

### get annual revenue 

In [11]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import numpy as np

url = 'https://finance.yahoo.com/quote/MSFT/financials?p=MSFT'
response = requests.get(url)
page = response.text
soup = BeautifulSoup(page, 'lxml')

tag = soup.find('span', string = 'Operating Expense')
print(tag.findNext('span').findNext('span').findNext('span').findNext('span'))

<span data-reactid="130">33,057,000</span>


In [12]:
def get_revenue(row):
    ''' 
    This method scrapes Yahoo Finance for EPS data on Jan 3rd 2020
    '''
    url = 'https://finance.yahoo.com/quote/{stock}/financials?p={stock}'.format(stock = row)
    response = requests.get(url)
    page = response.text
    soup = BeautifulSoup(page, 'lxml')
    try:
        tag = soup.find('span', string = 'Total Revenue')
        rev_2017 = tag.findNext('span').findNext('span').findNext('span').findNext('span')
        print(rev_2017.text)
        return rev_2017.text
    except:
        print(np.NaN)
        return np.NaN

def get_profit(row):
    '''
    this method returns 2017 reported gross profit for a stock from yahoo finance
    '''
    url = 'https://finance.yahoo.com/quote/{stock}/financials?p={stock}'.format(stock = row)
    response = requests.get(url)
    page = response.text
    soup = BeautifulSoup(page, 'lxml')
    try:
        tag = soup.find('span', string = 'Gross Profit')
        prof_2017 = tag.findNext('span').findNext('span').findNext('span').findNext('span')
        print(prof_2017.text)
        return prof_2017.text
    except:
        print(np.NaN)
        return np.NaN
    
def get_operating_expense(row):
    url = 'https://finance.yahoo.com/quote/{stock}/financials?p={stock}'.format(stock = row)
    response = requests.get(url)
    page = response.text
    soup = BeautifulSoup(page, 'lxml')
    try:
        tag = soup.find('span', string = 'Operating Expense')
        exp_2017 = tag.findNext('span').findNext('span').findNext('span').findNext('span')
        print(exp_2017.text)
        return exp_2017.text
    except:
        print(np.NaN)
        return np.NaN

In [13]:
df = pd.read_pickle('PE_Ratio_Data.pkl')

In [14]:
%%time
df['2017_Revenue'] = df['Ticker'].apply(get_revenue)


1,193,625
44,405
284,337
16,600
286,752
646,069
31,657,000
296,500
235,429
405,232
1,767,600
3,383,708
27,390,000
28,216,000
837
3,492,690
593,749
5,453,600
86,264
385,166
65,402
2,836,316
124,901
250,262
4,177
13,481
36,765,478
1,948,800
383,414
7,995
0
nan
1,024,191
1,683
60,501
588,287
7,017,000
nan
3,505,100
1,560,258
571
400,688
56,112
16,213,000
22,761
7,301,505
4,315,502
1,809,800
11,652,000
nan
0
3,735
425,500
230,574
1,124,511
3,551,100
129,829
1,473,413
903,374
0
6,424,655
3,525,100
109,529
1,503,778
2,408,200
15,940,700
1,419,300
7,719,400
3,382,200
252,012
2,262,000
Operating Expense
1,806,342
38,524,000
6,428,000
89,912
383,337
110,855,000
110,855,000
101,739
333,333
131,607
9,326,570
942,213
876,737
19,494,000
0
609,946
149,493
177,866,000
616,326
295,402
nan
5,079,200
2,805,691
3,867,155
1,533,680
0
3,686,345
349,643
176,842
113,420
7,927,400
1,684,265
1,095,250
54,515
0
2,747,920
90,039,400
70,074
9,998,000
5,887,000
1,005,437
0
1,009,591
1,326,173
255,202
356,216
143,8

7,304,000
333,633
117,559
584,933
268,130
1,375,308
874,448
67,693
769,030
342,029
nan
2,211,000
nan
8,570,000
138,580
178,638
619,793
1,236,915
4,060,000
456,492
297,141
2,066,982
15,478,000
10,551,000
57,521
223,986
4,706,900
120,535
93,707
25,803,000
nan
441,265
762,173
132,780
1,305,652
5,396,175
41,819
0
521,290
31,176
177,026
2,648,000
nan
10,629,000
2,279,099
1,939,362
20,252,393
766,869
1,029,520
333,034
29,864
2,097,900
9,714,000
6,313,690
8,977,726
40,156
1,248,424
12,508,000
1,921,507
196,290
nan
1,923
1,194,576
433,823
45,620
6,829,600
960,951
245,749
849,350
1,520,500
390,426
1,744,756
9,318,275
6,384,000
6,869,000
1,086,211
437,838
2,911,883
19,456,400
93,468
69,064
769,472
48,116
269,077
286,630
6,444,900
1,094,312
79,630
44,724
1,761,600
59,466
39,936
1,783,359
2,443,066
232,605
718,967
12,616
12,856
111,196
2,791,000
324,394
1,412,900
42,020
442,520
12,029,312
967,044
3,017,011
2,371,569
1,635,653
411,344
5,465,683
2,356,684
0
335,098
26,388
574,173
25,275
10,767,863
6

192,803
nan
518,504
1,081,454
23,470,967
22,245,500
12,586,000
2,787,979
108,524
5,157,000
2,371,900
737,405
1,004,900
6,939,600
159,335
903,221
474,927
812,052
nan
7,830,436
43,730,000
455,469
3,775,900
673,132
269,888
52,516
0
nan
7,686,999
341,700
560,873
15,264,000
nan
712,468
799,300
102,814
3,919,000
1,187,038
1,164,248
3,468,858
1,176,867
1,755,700
1,154
31,046
63,156
2,581,891
370,171
0
843,741
286,166
1,309,600
1,849,317
1,060,810
nan
1,342,532
592,884
11,074,481
113,196
211,698
536,034
683,679
107,804
11,225,806
1,450,448
10,000
1,410,997
575,991
2,987,994
3,362,200
4,368,428
895,558
1,215,294
1,105,009
604,296
598,118
237,162,000
7,539
747,954
2,090,041
1,420,490
40,653,000
1,221,179
932,169
95,972
541,500
46,829
97,697
46,219
360,917
4,390,500
4,106
294,852
735,478
175,159
857,348
898,500
1,102,924
65,450,000
391,662
1,396,742
Total Expenses
125,668
36,372
9,123,000
nan
669,132
5,729,000
147,235
nan
751,086
5,772,363
59,851
214,150
534,914
73,545
41,779
287,840
74,490
31,72

3,616,014
8,158
123,560
21,852,000
504,042
24,147,161
1,555,385
104,093
1,047,265
41,414
3,045,446
1,907,218
93,980,000
1,879,400
771,753
2,745,967
34,574
2,084,000
165,083
35,597
698,100
2,668,200
217,364
865
1,807,476
1,114,788
487,096
1,767,161
3,304,200
3,881,756
1,157,144
nan
0
118,214,000
711,857
500,343,000
1,169,092
463,852
758,122
325,078
184,340
14,485,000
2,309,078
175,146
605,884
4,341,955
1,456,700
3,704
4,720,895
380,506
1,055,765
7,648,500
1,306,911
573,163
3,466,807
1,445,400
nan
87,121,000
4,299,105
2,116,737
374,625
nan
7,679,021
68,163
1,599,100
187,939
807,831
105,331
19,093,000
67,895
5,524,300
8,041,000
14,859,700
133,785
1,250,548
639,239
170,201
48,589
162,667
4,891,000
7,144,000
5,878,000
519,495
2,199,800
3,722,000
430,492
1,076,794
1,076,794
7,824,100
2,609,000
6,389
65,663
5,307,000
9,821
125,717
927,401
167,926
0
23,432
861,390
CPU times: user 4min 38s, sys: 20.3 s, total: 4min 58s
Wall time: 1h 9min 12s


In [15]:
%%time
df['2017_Profit'] = df['Ticker'].apply(get_profit)

520,281
nan
nan
-707.912
229,376
304,839
15,656,000
225,329
182,111
123,397
276,500
3,124,774
15,053,000
21,176,000
nan
2,083,842
495,168
521,300
37,613
167,840
3,395
nan
111,841
172,648
3,175
nan
11,030,492
656,400
141,341
-2,664
nan
nan
571,905
476
nan
452,573
4,516,000
nan
1,481,200
800,792
554
106,095
27,475
1,408,000
-6,404
6,291,014
3,419,766
847,606
1,830,000
nan
nan
nan
123,500
144,704
798,902
3,096,900
8,930
1,116,947
335,737
nan
nan
449,000
nan
407,973
1,070,700
13,772,700
803,400
3,449,500
1,220,800
75,679
1,131,000
nan
782,161
nan
nan
Operating Expense
91,821
65,272,000
65,272,000
62,333
226,742
109,804
6,290,915
242,348
275,776
11,951,000
Operating Expense
426,780
nan
40,683,000
nan
187,733
nan
3,222,800
1,464,615
1,359,499
632,954
nan
318,799
176,169
97,810
86,056
1,571,000
nan
945,086
27,049
Operating Expense
644,093
nan
nan
3,909,000
2,028,000
651,058
nan
361,348
333,518
nan
81,560
88,520
112,140
460,516
88,186,000
737,695
6,532,000
166,280
865,102
2,614,000
19,761
522,

nan
774,625
nan
576,505
59,568
2,981,100
476,876
-9,903
12,499,000
179,248
194,393
1,241,860
9,987
2,990,000
156,142
30,376
181,363
3,592,000
33,500
nan
nan
128,097
20,780
119,809
nan
3,751,700
-2,639
313,949
1,638,436
144,030
121,565
nan
139,245
251,343
6,272,000
nan
258,111
124,175
142,739
62,850
nan
1,340,000
397,516
1,146,288
187,339
33,641
nan
36,799
nan
585,142
880,318
111,820
nan
884,100
730,613
5,475,166
8,684
80,769
318,575
199,833
854,014
2,864,796
26,288
150,972
4,092,094
3,881,266
192,153
544,418
2,165,688
190,297
142,521
282,827
98,709
nan
nan
829,936
829,936
113,018
1,463,768
582,500
332,029
nan
4,350,000
387,787
1,084,627
57,999
nan
12,640
nan
30,390
678,015
175,046
141,466
7,706,490
1,964,895
641,542
109,040
nan
141,923
nan
519,911
nan
147,231
172,599
nan
nan
6,781,211
159,187
-1,117
296,269
nan
12,393
nan
324,009
42,871
7,004
22,371
nan
-4,411
nan
2,190,000
661,116
454,191
9,129
129,617
37,215
nan
4,558,560
179,394
446,264
17,830
1,506,872
3,323,147
662,171
595,500
1,2

95,860
324,026
59,781
112,429
277,169
393,551
752,725
39,069,000
nan
4,807
1,826,396
376,286
nan
129,585
4,629,000
421,781
386,021
nan
1,806,445
nan
365,524
2,115,245
848,198
11,920
2,200,000
464,080
470,700
51,096,000
9,339,000
210,940
Operating Expense
nan
nan
3,072,100
331,203
231,034
9,278
311,600
87,937
nan
nan
547,472
1,496,200
1,470,800
227,359
722,061
342,000
nan
nan
5,022,000
954,100
340,393
nan
657,707
478,900
3,995,000
nan
1,702,000
nan
408,056
516,244
75,595
223,305
6,553,000
900,851
7,099,000
Operating Expense
nan
nan
2,291,100
542,039
46,238
nan
1,142,193
867,900
1,247,000
70,697
nan
600,513
1,624,884
nan
1,125,200
nan
88,436
168,479
320,420
396,276
nan
nan
nan
nan
nan
564,000
564,000
2,713,800
2,713,800
261,374
532,269
2,824,000
2,824,000
1,240,000
348,869
215,157
166,033
135,736
16,801,100
87,799
25,930
880,326
nan
130,978
4,982,000
145,012
nan
nan
12,298
143,788
1,516,086
1,656,300
nan
208,927
348,866
281,900
-5,483
3,484,600
253,419
696,000
13,289,000
40,952
273,143
2

In [16]:
%%time
df['2017_Operating_Expense'] = df['Ticker'].apply(get_operating_expense)

nan
nan
nan
12,592
259,514
358,812
8,422,000
nan
nan
48,346
nan
2,536,582
13,327,000
11,257,000
28,673
1,997,401
338,031
5,349,400
nan
nan
nan
2,398,434
nan
140,874
67,359
nan
6,397,883
441,700
151,164
36,018
nan
nan
487,265
72,899
nan
347,724
3,207,000
nan
nan
634,484
88,480
90,860
32,306
691,000
nan
4,122,919
3,072,499
618,152
334,000
nan
nan
36,414
81,600
125,304
491,792
2,292,900
15,024
763,336
nan
nan
nan
nan
nan
145,551
582,500
14,313,900
574,300
nan
567,400
nan
nan
nan
740,244
nan
nan
nan
78,677
36,390,000
36,390,000
69,986
259,999
128,003
nan
192,642
nan
2,362,000
1,738
nan
nan
nan
nan
163,302
nan
nan
707,980
nan
524,395
nan
287,930
135,692
68,578
40,327
nan
nan
554,358
nan
41,134
nan
nan
nan
nan
1,181,000
421,194
nan
262,398
nan
nan
43,982
nan
143,951
nan
26,842,000
563,105
2,664,000
79,118
541,375
nan
23,502
193,352
nan
nan
807,382
nan
125,773
nan
nan
1,993,000
nan
693,000
nan
nan
nan
98,715
nan
101,288
nan
60,046
2,168,200
2,783,200
nan
3,659,551
nan
2,782
634,823
nan
380,00

nan
494,088
23,845
2,277,800
1,311,518
nan
331,553
nan
40,299
nan
nan
nan
nan
431,286
nan
nan
nan
38,087
nan
nan
nan
nan
nan
nan
nan
1,196,032
5,039,000
nan
nan
2,027,000
10,153
nan
323,340
nan
nan
nan
815,260
nan
639,812
264,047
nan
nan
1,443,700
45,290,000
nan
4,457
222,000
nan
nan
nan
nan
nan
565,998
nan
14,433
103,016
54,496
nan
nan
nan
18,868
nan
355,778
248,837
2,947,000
376,751
nan
nan
222,000
1,299,000
121,993
5,002,000
111,641
nan
1,474,984
nan
81,998
nan
nan
357,347
87,281
nan
2,191,000
nan
nan
nan
nan
nan
nan
nan
nan
1,477,900
131,489
2,318,000
173,831
40,527
nan
nan
nan
nan
nan
303,152
nan
191,927
nan
nan
266,400
nan
nan
595,000
833.545
603,733
nan
1,634,000
371,000
1,257,834
nan
34,059
2,300,000
89,273
184,529
28,181
3,130,000
1,004,231
nan
1,656,500
1,270,000
nan
19,244
265,923
123,692
3,514,837
1,401,400
199,224
17,621
nan
nan
72,815
60,696
nan
nan
32,881
145,693
nan
147,500
77,259
nan
nan
nan
nan
nan
1,350,800
149,866
9,653
1,568,000
74,549
nan
nan
41,701
133,274
111,12

nan
nan
nan
nan
2,688,000
nan
998,000
103,900
143,468
nan
nan
nan
1,359,600
177,809
nan
nan
nan
231,500
nan
136,709
nan
318,083
nan
nan
227,292
nan
nan
nan
nan
282,900
nan
nan
nan
391,957
nan
nan
184,230
nan
nan
441,600
339,771
nan
64,062
462,556
43,268
1,041,975
nan
nan
3,128
nan
nan
nan
nan
41,863
nan
1,174,303
744,262
99,536
107,038
nan
nan
204,900
nan
nan
482,928
nan
nan
nan
nan
nan
nan
69,810
nan
32,879
418,977
10,141
nan
nan
656,000
nan
nan
nan
158,372
nan
202,899
18,243,000
120,977
294,245
nan
565,727
nan
nan
nan
nan
nan
1,805,583
28,415
42,072
498,326
59,658
61,334
22,563
nan
nan
nan
nan
nan
109,986
818,100
100,993
275,161
nan
895,045
47,122
nan
129,570
1,139,700
379,000
454,800
239,000
1,360,000
nan
nan
nan
nan
nan
251,000
556,909
nan
nan
nan
nan
nan
nan
57,426
868,000
7,212,000
nan
31,802,000
506,250
68,000
nan
nan
1,446,300
nan
134,684
nan
nan
48,204
3,043,954
nan
nan
nan
nan
nan
nan
10,242
127,265
915,615
27,508
36,599
nan
84,466
nan
nan
26,511
nan
nan
1,229,046
nan
948,000

In [17]:
df

Company Ticker Open_Price_2018 Open_Price_2020  \
0            1-800 FLOWERS COM   FLWS           11.15           13.96   
1     1ST CONSTITUTION BANCORP   FCCY            18.2           22.35   
2              1ST SOURCE CORP   SRCE           49.89           51.25   
3       22ND CENTURY GROUP INC   XXII            3.12            1.13   
4                       2U INC   TWOU           64.26           23.12   
...                        ...    ...             ...             ...   
3007                ZUMIEZ INC   ZUMZ            23.1           33.75   
3008                     ZUORA    ZUO             NaN           14.56   
3009   ZYNERBA PHARMACEUTICALS   ZYNE           12.81             5.8   
3010                     ZYNEX   ZYXI            3.24            7.81   
3011                 ZYNGA INC   ZNGA            3.92            6.12   

      Percent Increase  EPS  P/E Ratio 2017_Revenue 2017_Profit  \
0            25.201794  0.0        NaN    1,193,625     520,281   
1            22.802198  0.0        NaN       44,405         NaN   
2             2.725997  2.6  19.188462      284,337         NaN   
3           -63.782051  0.0        NaN       16,600    -707.912   
4           -64.021164  0.0        NaN      286,752     229,376   
...                ...  ...        ...          ...         ...   
3007         46.103896  1.1  21.000000      927,401         NaN   
3008               NaN -1.1        NaN      167,926      88,020   
3009        -54.722873 -2.5  -5.124000            0         NaN   
3010        141.049383  0.0        NaN       23,432         NaN   
3011         56.122449  0.0        NaN      861,390     602,419   

     2017_Operating_Expense  
0                       NaN  
1                       NaN  
2                       NaN  
3                    12,592  
4                   259,514  
...                     ...  
3007                    NaN  
3008                    NaN  
3009                    NaN  
3010                    NaN  
3011                576,695  

[3012 rows x 10 columns]

In [7]:
df.to_pickle('Rev_Prof_OE_data.pkl')



NameError: name 'df' is not defined

In [8]:
df2 = pd.read_pickle("Rev_Prof_OE_data.pkl")
df2

NameError: name 'pd' is not defined

In [39]:
float(df2['2017_Revenue'][0].replace(',',''))

1193625.0

In [43]:
def convert_to_float(row):
    try:
        return float(row.replace(',',''))
    except:
        return np.NaN

In [45]:
df2['2017_Revenue'] = df2['2017_Revenue'].apply(convert_to_float)

In [46]:
df2['2017_Profit'] = df2['2017_Profit'].apply(convert_to_float)
df2['2017_Operating_Expense'] = df2['2017_Operating_Expense'].apply(convert_to_float)

In [49]:
df2.corr()

Percent Increase       EPS  P/E Ratio  2017_Revenue  \
Percent Increase                1.000000 -0.017103   0.010963     -0.017757   
EPS                            -0.017103  1.000000   0.142006      0.004302   
P/E Ratio                       0.010963  0.142006   1.000000      0.042783   
2017_Revenue                   -0.017757  0.004302   0.042783      1.000000   
2017_Profit                    -0.004778  0.005652   0.058368      0.854920   
2017_Operating_Expense         -0.003998  0.006348   0.036860      0.887035   

                        2017_Profit  2017_Operating_Expense  
Percent Increase          -0.004778               -0.003998  
EPS                        0.005652                0.006348  
P/E Ratio                  0.058368                0.036860  
2017_Revenue               0.854920                0.887035  
2017_Profit                1.000000                0.959695  
2017_Operating_Expense     0.959695                1.000000

In [3]:
def get_assets(row):
    '''
    this method returns 2017 reported gross profit for a stock from yahoo finance
    '''
    url = 'https://finance.yahoo.com/quote/{stock}/balance-sheet?p={stock}'.format(stock = row)
    response = requests.get(url)
    page = response.text
    soup = BeautifulSoup(page, 'lxml')
    try:
        tag = soup.find('span', string = 'Total Assets')
        prof_2017 = tag.findNext('span').findNext('span').findNext('span')
        print(prof_2017.text)
        return prof_2017.text
    except:
        print(np.NaN)
        return np.NaN

def get_liabilities(row):
    '''
    this method returns 2017 reported gross profit for a stock from yahoo finance
    '''
    url = 'https://finance.yahoo.com/quote/{stock}/balance-sheet?p={stock}'.format(stock = row)
    response = requests.get(url)
    page = response.text
    soup = BeautifulSoup(page, 'lxml')
    try:
        tag = soup.find('span', string = 'Total Liabilities Net Minority Interest')
        prof_2017 = tag.findNext('span').findNext('span').findNext('span')
        print(prof_2017.text)
        return prof_2017.text
    except:
        print(np.NaN)
        return np.NaN


In [4]:
df2['Total_Assets'] = df2['Ticker'].apply(get_assets)
df2['Total_Liabilities'] = df2['Ticker'].apply(get_liabilities)

NameError: name 'df2' is not defined

In [5]:
df2.to_pickle('Assets_Liabilities_data.pkl')

NameError: name 'df2' is not defined

In [6]:
df3 = pd.read_pickle('Assets_Liabilities_data.pkl')

NameError: name 'pd' is not defined

In [55]:
url = 'https://finance.yahoo.com/quote/AAPL/balance-sheet?p=AAPL'
response = requests.get(url)
page = response.text
soup = BeautifulSoup(page, 'lxml')

tags = soup.find_all('span')
for i, tag in enumerate(tags):
    print(i, tag.text)

0 No matching results for ''
1 Tip: Try a valid symbol or a specific company name for relevant results
2 Cancel
3 Summary
4 Company Outlook
5 Statistics
6 Historical Data
7 Profile
8 Financials
9 Analysis
10 Options
11 Holders
12 Sustainability
13 NasdaqGS - NasdaqGS Real Time Price. Currency in USD
14 Show:
15 Show
16 Income Statement
17 Balance Sheet
18 Cash Flow
19 Annual
20 Quarterly
21 Subscribe to Premium to download historical dataDownload
22 Subscribe to Premium to download historical data
23 Download
24 Balance Sheet
25 Expand All
26 All numbers in thousands
27 All numbers in thousands
28 Get access to 40+ years of historical data with Yahoo Finance Premium.
29 Learn more
30 Breakdown
31 9/30/2019
32 9/30/2018
33 9/30/2017
34 9/30/2016
35 Total Assets
36 338,516,000
37 365,725,000
38 375,319,000
39 321,686,000
40 Total Liabilities Net Minority Interest
41 248,028,000
42 258,578,000
43 241,272,000
44 193,437,000
45 Total Equity Gross Minority Interest
46 90,488,000
47 107,147,0


### 3. Price to earnings ratio to growth ratio (PEG)
This helps you understand the P/E ratio a little better. It’s calculated by dividing the P/E ratio by the company’s projected growth
 in earnings.

Example – A stock with a P/E of 30 and projected earnings growth next year of 15% would have a PEG of 2 (30 divided by 15). A stock with a P/E of 30 but projected earnings growth of 30% will have PEG of 1 (30 divided by 30).

The PEG can tell you whether a stock may or may not be a good value. The lower the number, the less you have to pay to get in on the company’s expected future earnings growth.

In [35]:
#scrape finbox for PEG ratios

url = 'https://www.morningstar.com/stocks/xnas/aapl/valuation'
response = requests.get(url)
page = response.text
soup = BeautifulSoup(page, 'lxml')
for i,tag in enumerate(soup.find_all('span')):
    print(i, tag.text)


# scrape projected growth in eanings and then divide it by the P/E ratio

0 
1 
2 
		Get 14 Days Free
	
3 
		Sign Up
	
4 
		Sign In
	
5 Learn
6 Learn
7 Start Investing 
8 Start Investing
9 Investing Classroom 
10 Investing Classroom
11 Investing Glossary 
12 Investing Glossary
13 Plan
14 Plan
15 Tax Planning 
16 Tax Planning
17 Personal Finance 
18 Personal Finance
19 Save for College 
20 Save for College
21 Save for Retirement 
22 Save for Retirement
23 Invest in Retirement 
24 Invest in Retirement
25 Research
26 Research
27 Mutual Funds 
28 Mutual Funds
29 Stocks 
30 Stocks
31 ETFs 
32 ETFs
33 Bonds 
34 Bonds
35 Best Investments 
36 Best Investments
37 Portfolio
38 Markets
39 
					Learn
					
40 
							Start Investing
						 
41 
							Investing Classroom
						 
42 
							Investing Glossary
						 
43 
					Plan
					
44 
							Tax Planning
						 
45 
							Personal Finance
						 
46 
							Save for College
						 
47 
							Save for Retirement
						 
48 
							Invest in Retirement
						 
49 
					Research
					
50 
							Mutual Funds
						


### 4. Price to book value ratio (P/B)
This compares the value the market puts on a company with the value the company has stated in its financial books. It’s calculated by dividing the current price per share by the book value per share. The book value is the current equity of a company, as listed in the annual report.

Most of the time, the lower the P/B is, the better. That’s because you’re paying less for more book value.

If you’re looking for a well-priced stock with reasonable growth potential, you may want to use a low P/B as a tool to identify possible stock picks.




### 5. Dividend payout ratio (DPR)
This measures what a company pays out to investors in dividends compared to what the stock is earning. It’s calculated by dividing the annual dividends per share by the EPS.

Example – If a company paid out $1 per share in dividends and had an EPS of $3, the DPR would be 33% (1 divided by 3).

The DPR can give you an idea of how well a company’s earnings support the dividend payments. More mature companies will typically have a higher DPR. They believe that paying more in dividends is the best use of their profits for the firm and its shareholders. Since growing companies are likely to have less or no earnings to pay out dividends, their DPR would tend to be low or zero.

### 6. Dividend yield
This measures the return on a dividend as a percentage of the stock price. It’s calculated by dividing the annual dividend
 per share by the price per share.

Example – 2 stocks each pay an annual dividend of $1 per share. Company A’s stock is trading at $40 a share, but Company B’s stock is trading at $20 a share. Company A has a dividend yield
 of 2.5% (1 divided by 40), while Company B’s is 5% (1 divided by 20).

The dividend yield
 can tell you how much cash flow
 you’re getting for your money, all other things being equal.

CAUTION
Indicators can help you assess the value of a stock and its growth potential. But there are many other factors affecting stock prices that can’t be easily measured.